### DISCUSSION TITLE
Calling Snippets Directly

### DISCUSSION THREAD

**Discussion by ThunderSloth** [L3MON4D3/LuaSnip#1081] [Q&A] [answered]
Is there a way to trigger a snippet directly with lua using some sort of identifier? 

One use case for this would be the ability to have modular snippets that could reference each-other. Right now, I achieve this with a lua module containing dynamic nodes that return whatever nodes I like to reuse. But I think it would be much more straightforward if I could just call another snippet directly.

**Comment by L3MON4D3** [comment]
You can use `snip_expand`(check `DOC.md`), you only have to store the snippets somewhere separate from luasnips snippet-collection since retrieving them from there is not really efficient.

But tbh, reusing dynamicNodes for this purpose sounds more straightforward than expanding a whole other snippet.

**Reply by ThunderSloth** [reply]
Okay, so I'm running into issues with dynamic nodes calling each other. 

Basically I have one dynamic node function called `paste`, when called returns each line of `LS_SELECT_RAW` as an insert node with the default text equal to that line. I use it with some autosnippets on various brackets in order to easily surround, and if the select buffer is empty I have it set to return one insert node with empty default text. 

The second dynamic node function (which is for LaTeX) I have called `smartpairs"`returns two function nodes that surround an insert node. The function nodes return the left and right sides of a set of parenthesis respectively, however if the insert node meets certain conditions, they return `"\left("` and `"\right)"` instead.

Both of these functions work perfectly fine on their own but I'm trying to chain them together, having `smartpairs` call `paste` and I'm running into an error about attempting to index a nil value. The reason is because it seems there is no way to access the .env variable when a dynamic node function is called within another dynamic node function (at least the way I am currently trying.) I have attempted using parent to access the env (which works fine if not called from another dynamic node.) Makes sense that this doesn't work because things get tricky with dynamic/sn and parents. So to circumvent this I tried `ls.session.current_nodes[vim.api.nvim_get_current_buf()]` instead, but this didn't work either. After more debugging I found that the entire ls.sessions table to be nil.

**tldr: when a dynamic node function calls another dynamic code function I cant access env in order to grab selected text**

The `smartpairs` function that gets returned in a table:

```
	smart_pairs = function(args, parent, old_state, user_args)
		local nodes, LHS, RHS = unpack(user_args)
		local use_LR = function(text)
			return text[1][1]:match("frac") ~= nil or text[1][1]:match("[_^]") ~= nil
		end
		return sn(1, {
			f(function(text)
				return (use_LR(text) and "\\left" or "") .. LHS
			end, { 1 }),
			d(1, utils.paste, {}, {user_args = {nodes}}), -- replace this line with "nodes," and it works just fine
			f(function(text)
				return (use_LR(text) and "\\right" or "") .. RHS
			end, { 1 }),
		})
	end,
```

The `paste` function that gets returned in a table:
```
	paste = function(args, parent, old_state, user_args)
		print(require("luasnip.sessions"))
		local result, env = {}, ls.session.current_nodes[vim.api.nvim_get_current_buf()].env
		for j, v in ipairs(env.LS_SELECT_RAW) do
			table.insert(result, i(j, v))
			if j ~= #env.LS_SELECT_RAW then
				table.insert(result, t({"", ""}))
			end
		end
		return sn(nil, #result ~= 0 and result or {user_args and user_args.default or i(1)})
	end,
```
The snippet where the calls originate:
```
	s( -- parenthesis
		{trig = "(", snippetType = "autosnippet", priority = 2000},
		{d(1, tex.smart_pairs, {}, {user_args = {{{i(1, "")}, "(", ")"}}}),}
	),
```
I imagine I could sacrifice some DRY and flatten these two functions into one, but I'm curious if you can tell me what's going on here. Thanks!

EDIT: Or now that I'm thinking about it I could make a copy of the selected text table and pass it through to the paste function  as a user_arg in case it can't access the parent.

**Reply by L3MON4D3** [reply, marked as answer]
Mhmm, at the very least, that's a very curious way of accessing `env` :D
Try `local env = parent.snippet.env`, that might do it
And damn, what timing, I just checked my mail xD

**Reply by ThunderSloth** [reply]
It works! I think at some point I accidentally switched from using `parent.snippet.env` to `parent.env` and `ls.session.current_nodes[vim.api.nvim_get_current_buf()].env` is what I saw in the external dynamic node update example on the wiki.
